aten/src/ATen/native/cpu/SoftMaxKernel.cpp

```diff
@@ -0,0 +1,282 @@
+            scalar_t* output_data = output_data_base + i * dim_size;
+            scalar_t tmp_sum = tmp_sum_scalar[j];
+            vec256::map(
+                [tmp_sum](Vec x) { return x - Vec(tmp_sum); },
+                output_data,
+                input_data,
+                dim_size);
+          }
+        }
+      },
+      ap);
+}
+
+template <typename scalar_t>
+inline void _vec_softmax_lastdim(
+    scalar_t* input_data_base,
+    scalar_t* output_data_base,
+    int64_t outer_size,
+    int64_t dim_size) {
+  using Vec = vec256::Vec256<scalar_t>;
+  int64_t grain_size = internal::TBB_GRAIN_SIZE / (16 * dim_size);
+  if (grain_size < 1)
+    grain_size = 1;
+
+  tbb::parallel_for(
+      tbb::blocked_range<int64_t>(0, outer_size, grain_size),
+      [&](const tbb::blocked_range<int64_t>& r) {
+        for (int64_t i = r.begin(); i < r.end(); i++) {
+          scalar_t* input_data = input_data_base + i * dim_size;
+          scalar_t* output_data = output_data_base + i * dim_size;
+          scalar_t max_input = vec256::reduce_all<scalar_t>(
+              [](Vec& x, Vec& y) { return vec256::max(x, y); },
+              input_data,
+              dim_size);
+          vec256::map(
+              [max_input](Vec x) { return (x - Vec(max_input)).exp(); },
+              output_data,
+              input_data,
+              dim_size);
+          scalar_t tmp_sum = vec256::reduce_all<scalar_t>(
+              [](Vec x, Vec y) { return x + y; }, output_data, dim_size);
+          tmp_sum = 1 / tmp_sum;
+          vec256::map(
+              [tmp_sum](Vec x) { return x * Vec(tmp_sum); },
+              output_data,
+              output_data,
+              dim_size);
+        }
+      },
+      ap);
+}
+
+template <typename scalar_t, bool log_softmax>
+inline void _vec_host_softmax_backward_lastdim(
+    scalar_t* grad_input_data_base,
+    scalar_t* grad_data_base,
+    scalar_t* output_data_base,
+    int64_t outer_size,
+    int64_t dim_size) {
+  using Vec = vec256::Vec256<scalar_t>;
+  int64_t grain_size = internal::TBB_GRAIN_SIZE / (16 * dim_size);
+  if (grain_size < 1)
+    grain_size = 1;
+
+  tbb::parallel_for(
+      tbb::blocked_range<int64_t>(0, outer_size, grain_size),
+      [&](const tbb::blocked_range<int64_t>& r) {
+        for (int64_t i = r.begin(); i < r.end(); i++) {
+          scalar_t* grad_input_data = grad_input_data_base + i * dim_size;
+          scalar_t* grad_data = grad_data_base + i * dim_size;
+          scalar_t* output_data = output_data_base + i * dim_size;
+          scalar_t sum;
+          if (log_softmax) {
+            sum = vec256::reduce_all<scalar_t>(
+                [](Vec& x, Vec& y) { return x + y; }, grad_data, dim_size);
+          } else {
+            sum = vec256::map2_reduce_all<scalar_t>(
+                [](Vec x, Vec y) { return x * y; },
+                [](Vec x, Vec y) { return x + y; },
+                grad_data,
+                output_data,
+                dim_size);
```

Comment:
That seems different than the grad formula we're using for CUDA I think (which is using a plain reduce_all over grad_data for both softmax and log softmax). Where is this coming from?